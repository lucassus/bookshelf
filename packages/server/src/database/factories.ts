import { getManager } from "typeorm";

import { Author } from "./entity/Author";
import { Avatar } from "./entity/Avatar";
import { Book } from "./entity/Book";
import { BookCopy } from "./entity/BookCopy";
import { User } from "./entity/User";

export const createAvatar = (attributes: Partial<Avatar> = {}) => {
  const manager = getManager();

  return manager.save(
    manager.create(Avatar, {
      imagePath: "/images/avatars/w13.png",
      color: "yellow",
      ...attributes
    })
  );
};

export const createUser = async (
  attributes: Partial<User> = { name: "Alice" }
) => {
  const manager = getManager();

  const { ...userAttributes } = attributes;

  if (userAttributes.name && userAttributes.email === undefined) {
    userAttributes.email = `${userAttributes.name.toLocaleLowerCase()}@email.com`;
  }

  if (userAttributes.avatarId === undefined) {
    const avatar = await createAvatar();
    userAttributes.avatarId = avatar.id;
  }

  const user = manager.create(User, userAttributes);

  return manager.save(user);
};

export const createAuthor = (attributes: Partial<Author> = {}) => {
  const manager = getManager();

  return manager.save(
    manager.create(Author, {
      name: "Andrzej Sapkowski",
      photoPath: "/images/book-authors/andrzej-sapkowski.jpg",
      ...attributes
    })
  );
};

export const createBook = async (
  attributes: Partial<Book> & { authorAttributes?: Partial<Author> } = {}
) => {
  const manager = getManager();

  const { authorAttributes, ...bookAttributes } = attributes;

  if (bookAttributes.authorId === undefined) {
    const author = await createAuthor(authorAttributes);
    bookAttributes.authorId = author.id;
  }

  return manager.save(
    manager.create(Book, {
      title: "Baptism of fire",
      coverPath: "/images/book-covers/witcher3.jpg",
      ...bookAttributes
    })
  );
};

export const createBookCopy = async (
  attributes: Partial<BookCopy> & { bookAttributes?: Partial<Book> } & {
    ownerAttributes?: Partial<User>;
  } = {}
) => {
  const manager = getManager();

  const { bookAttributes, ownerAttributes, ...bookCopyAttributes } = attributes;

  if (bookCopyAttributes.bookId === undefined) {
    const book = await createBook(bookAttributes);
    bookCopyAttributes.bookId = book.id;
  }

  if (bookCopyAttributes.ownerId === undefined) {
    const user = await createUser(ownerAttributes);
    bookCopyAttributes.ownerId = user.id;
  }

  return manager.save(manager.create(BookCopy, bookCopyAttributes));
};
