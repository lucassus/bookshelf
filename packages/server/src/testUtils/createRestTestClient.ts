import express from "express";
import request from "supertest";

import { User } from "../database/entity/User";
import { routes } from "../rest";

export function createRestTestClient({
  currentUser
}: { currentUser?: User } = {}) {
  const app = express();

  if (currentUser) {
    app.use((req, res, next) => {
      (req as any).user = currentUser;
      next();
    });
  }

  app.use("/", routes);

  return request(app);
}
