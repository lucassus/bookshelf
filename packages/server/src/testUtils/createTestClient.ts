import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";

import { buildContext, Context } from "../graphql/context";
import { rootSchema } from "../graphql/rootSchema";

export function createTestClient(
  contextExtra: Partial<Context> = {}
): ApolloServerTestClient {
  const server = new ApolloServer({
    schema: rootSchema,
    context: buildContext(contextExtra)
  });

  return createApolloTestClient(server);
}
