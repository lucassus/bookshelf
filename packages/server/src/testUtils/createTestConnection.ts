import path from "path";
import { createConnection } from "typeorm";

export const createTestConnection = () =>
  createConnection({
    type: "sqlite",
    database: ":memory:",
    dropSchema: true,
    entities: [path.join(__dirname, "../entity/**/*.ts")],
    synchronize: true,
    logging: false
  });
