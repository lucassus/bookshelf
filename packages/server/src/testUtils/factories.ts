import faker from "faker";
import { getManager, ObjectType, DeepPartial } from "typeorm";

import { hashPassword } from "../common/authentication";
import { titleizeSentence } from "../common/strings";
import { Author } from "../database/entity/Author";
import { Avatar, AVATAR_COLORS } from "../database/entity/Avatar";
import { Book } from "../database/entity/Book";
import { BookCopy } from "../database/entity/BookCopy";
import { User } from "../database/entity/User";

const AVATAR_IMAGES = [
  "/images/avatars/w13.png",
  "/images/avatars/m10.png",
  "/images/avatars/w2.png",
  "/images/avatars/m25.png"
];

const AUTHOR_PHOTOS = [
  "/images/book-authors/j-k-rowling.jpg",
  "/images/book-authors/james-s-a-corey.jpg",
  "/images/book-authors/andrzej-sapkowski.jpg"
];

const BOOK_COVERS = [
  "/images/book-covers/harry1.jpg",
  "/images/book-covers/harry2.jpg",
  "/images/book-covers/harry3.jpg",
  "/images/book-covers/harry4.jpg",
  "/images/book-covers/harry5.jpg",
  "/images/book-covers/harry6.jpg",
  "/images/book-covers/harry7.jpg",
  "/images/book-covers/expanse1.jpg",
  "/images/book-covers/expanse2.jpg",
  "/images/book-covers/expanse3.jpg",
  "/images/book-covers/expanse4.jpg",
  "/images/book-covers/expanse5.jpg",
  "/images/book-covers/expanse6.jpg",
  "/images/book-covers/expanse7.jpg",
  "/images/book-covers/expanse8.jpg",
  "/images/book-covers/witcher1.jpg",
  "/images/book-covers/witcher2.jpg",
  "/images/book-covers/witcher3.jpg",
  "/images/book-covers/witcher4.jpg",
  "/images/book-covers/witcher5.jpg"
];

function createEntity<Entity>(
  entityClass: ObjectType<Entity>,
  attributes: DeepPartial<Entity>
) {
  const manager = getManager();

  const entity = manager.create(entityClass, attributes);
  return manager.save(entity);
}

type CreateAvatarAttributes = Partial<Avatar>;

export function createAvatar(attributes: CreateAvatarAttributes = {}) {
  return createEntity(Avatar, {
    imagePath: faker.random.arrayElement(AVATAR_IMAGES),
    color: faker.random.arrayElement(AVATAR_COLORS),
    ...attributes
  });
}

type CreateUserAttributes = Partial<User> & {
  password?: string;
  avatarAttributes?: CreateAvatarAttributes;
};

export async function createUser(attributes: CreateUserAttributes = {}) {
  const { avatarAttributes, password, ...userAttributes } = attributes;

  if (userAttributes.name === undefined) {
    userAttributes.name = faker.name.findName();
  }

  if (userAttributes.email === undefined) {
    userAttributes.email = faker.internet.email();
  }

  if (userAttributes.info === undefined) {
    userAttributes.info = faker.lorem.sentence();
  }

  if (userAttributes.avatarId === undefined) {
    const avatar = await createAvatar();
    userAttributes.avatarId = avatar.id;
  }

  if (avatarAttributes) {
    const avatar = await createAvatar(avatarAttributes);
    userAttributes.avatarId = avatar.id;
  }

  return createEntity(User, {
    passwordHash: hashPassword(password || "password"),
    isAdmin: false,
    ...userAttributes
  });
}

export function createAuthor(attributes: Partial<Author> = {}) {
  return createEntity(Author, {
    name: faker.name.findName(),
    bio: faker.lorem.sentence(),
    photoPath: faker.random.arrayElement(AUTHOR_PHOTOS),
    ...attributes
  });
}

type CreateBookAttributes = Partial<Book> & {
  authorAttributes?: Partial<Author>;
};

export async function createBook(attributes: CreateBookAttributes = {}) {
  const { authorAttributes, ...bookAttributes } = attributes;

  if (bookAttributes.authorId === undefined) {
    const author = await createAuthor(authorAttributes);
    bookAttributes.authorId = author.id;
  }

  return createEntity(Book, {
    title: titleizeSentence(
      faker.lorem.words(faker.random.number({ min: 1, max: 4 }))
    ),
    description: faker.lorem.sentence(),
    coverPath: faker.random.arrayElement(BOOK_COVERS),
    favourite: faker.random.boolean(),
    ...bookAttributes
  });
}

export async function createBookCopy(
  attributes: Partial<BookCopy> & {
    bookAttributes?: CreateBookAttributes;
  } & {
    ownerAttributes?: CreateUserAttributes;
  } & { borrowerAttributes?: CreateUserAttributes } = {}
) {
  const {
    bookAttributes,
    ownerAttributes,
    borrowerAttributes,
    ...bookCopyAttributes
  } = attributes;

  if (bookCopyAttributes.bookId === undefined) {
    const book = await createBook(bookAttributes);
    bookCopyAttributes.bookId = book.id;
  }

  if (bookCopyAttributes.ownerId === undefined) {
    const owner = await createUser(ownerAttributes);
    bookCopyAttributes.ownerId = owner.id;
  }

  if (borrowerAttributes) {
    const borrower = await createUser(borrowerAttributes);
    bookCopyAttributes.borrowerId = borrower.id;
  }

  return createEntity(BookCopy, bookCopyAttributes);
}
