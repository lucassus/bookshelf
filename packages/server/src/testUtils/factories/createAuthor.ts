import faker from "faker";

import { Author } from "../../database/entity/Author";
import { createEntity } from "./createEntity";

const AUTHOR_PHOTOS = [
  "/images/book-authors/j-k-rowling.jpg",
  "/images/book-authors/james-s-a-corey.jpg",
  "/images/book-authors/andrzej-sapkowski.jpg"
];

export type CreateAuthorAttributes = Partial<Author>;

export function createAuthor(
  attributes: CreateAuthorAttributes = {}
): Promise<Author> {
  return createEntity(Author, {
    name: faker.name.findName(),
    bio: faker.lorem.sentence(),
    photoPath: faker.random.arrayElement(AUTHOR_PHOTOS),
    ...attributes
  });
}
