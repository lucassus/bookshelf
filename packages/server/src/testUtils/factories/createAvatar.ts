import faker from "faker";

import { Avatar } from "../../database/entity/Avatar";
import { createEntity } from "./createEntity";

const AVATAR_IMAGES = [
  "/images/avatars/w13.png",
  "/images/avatars/m10.png",
  "/images/avatars/w2.png",
  "/images/avatars/m25.png"
];

const AVATAR_COLORS = [
  "red",
  "green",
  "blue",
  "yellow",
  "magenta",
  "pink",
  "black"
];

export type CreateAvatarAttributes = Partial<Avatar>;

export function createAvatar(
  attributes: CreateAvatarAttributes = {}
): Promise<Avatar> {
  return createEntity(Avatar, {
    imagePath: faker.random.arrayElement(AVATAR_IMAGES),
    color: faker.random.arrayElement(AVATAR_COLORS),
    ...attributes
  });
}
