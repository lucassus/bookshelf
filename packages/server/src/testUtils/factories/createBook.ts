import faker from "faker";

import { titleizeSentence } from "../../common/strings";
import { Author } from "../../database/entity/Author";
import { Book } from "../../database/entity/Book";
import { createAuthor, CreateAuthorAttributes } from "./createAuthor";
import { createEntity } from "./createEntity";

const BOOK_COVERS = [
  "/images/book-covers/harry1.jpg",
  "/images/book-covers/harry2.jpg",
  "/images/book-covers/harry3.jpg",
  "/images/book-covers/harry4.jpg",
  "/images/book-covers/harry5.jpg",
  "/images/book-covers/harry6.jpg",
  "/images/book-covers/harry7.jpg",
  "/images/book-covers/expanse1.jpg",
  "/images/book-covers/expanse2.jpg",
  "/images/book-covers/expanse3.jpg",
  "/images/book-covers/expanse4.jpg",
  "/images/book-covers/expanse5.jpg",
  "/images/book-covers/expanse6.jpg",
  "/images/book-covers/expanse7.jpg",
  "/images/book-covers/expanse8.jpg",
  "/images/book-covers/witcher1.jpg",
  "/images/book-covers/witcher2.jpg",
  "/images/book-covers/witcher3.jpg",
  "/images/book-covers/witcher4.jpg",
  "/images/book-covers/witcher5.jpg"
];

export type CreateBookAttributes = Omit<Partial<Book>, "author"> & {
  author?: Author;
  authorAttributes?: CreateAuthorAttributes;
};

export async function createBook(
  attributes: CreateBookAttributes = {}
): Promise<Book> {
  const { author, authorAttributes, ...bookAttributes } = attributes;

  if (bookAttributes.authorId === undefined) {
    bookAttributes.authorId = author
      ? author.id
      : (await createAuthor(authorAttributes)).id;
  }

  return createEntity(Book, {
    title: titleizeSentence(
      faker.lorem.words(faker.random.number({ min: 1, max: 4 }))
    ),
    description: faker.lorem.sentence(),
    coverPath: faker.random.arrayElement(BOOK_COVERS),
    favourite: faker.random.boolean(),
    ...bookAttributes
  });
}
