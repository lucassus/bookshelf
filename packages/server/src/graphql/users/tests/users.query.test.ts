import { gql } from "apollo-server-express";

import { createTestClient } from "../../../testUtils/createTestClient";
import { createUser } from "../../../testUtils/factories";
import { Role } from "../../resolvers-types.generated";

describe("users query", () => {
  it("fetches users", async () => {
    // Given
    const user1 = await createUser({ name: "Alice" });
    const user2 = await createUser({ name: "Bob" });
    const user3 = await createUser({ name: "Luke" });

    // When
    const res = await createTestClient().query({
      query: gql`
        query {
          users {
            id
            name
            info
            avatar {
              image {
                path
                url
              }
              color
            }
          }
        }
      `
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).not.toBe(null);
    expect(res.data).toMatchObject({
      users: [
        {
          id: expect.any(String),
          name: user1.name,
          info: user1.info
        },
        {
          id: expect.any(String),
          name: user2.name,
          info: user2.info
        },
        {
          id: expect.any(String),
          name: user3.name,
          info: user3.info
        }
      ]
    });
  });

  describe("fetching protected users emails", () => {
    test.each<[undefined | Role, undefined | string]>([
      [undefined, "Unauthorized access! Please log in."],
      [Role.User, "Unauthorized access! Please log in as admin."],
      [Role.Admin, undefined]
    ])(
      "for `%s` role raises `%s` error",
      async (role, expectedErrorMessage) => {
        // Given
        const currentUser = role
          ? await createUser({ isAdmin: role === Role.Admin })
          : undefined;
        const otherUser = await createUser();

        // When
        const res = await createTestClient({ currentUser }).query({
          query: gql`
            query {
              users {
                id
                email
              }
            }
          `
        });

        if (expectedErrorMessage) {
          expect(res.errors).not.toBe(undefined);
          expect(res.errors![0].message).toBe(expectedErrorMessage);
        } else {
          expect(res.errors).toBe(undefined);
          expect(res.data).toMatchObject({
            users: [
              { id: expect.any(String), email: currentUser!.email },
              { id: expect.any(String), email: otherUser.email }
            ]
          });
        }
      }
    );
  });
});
