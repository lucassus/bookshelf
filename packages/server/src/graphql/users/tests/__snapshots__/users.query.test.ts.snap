// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users query 1`] = `
Array [
  Object {
    "avatar": Object {
      "color": "red",
      "image": Object {
        "path": "/images/avatars/w2.png",
        "url": "https://res.cloudinary.com/lucassus/image/upload/bookshelf/images/avatars/w2.png",
      },
    },
    "info": "Velit quaerat ab veniam.",
    "name": "Mariana Christiansen",
  },
  Object {
    "avatar": Object {
      "color": "yellow",
      "image": Object {
        "path": "/images/avatars/w13.png",
        "url": "https://res.cloudinary.com/lucassus/image/upload/bookshelf/images/avatars/w13.png",
      },
    },
    "info": "Exercitationem sequi est minus consequatur.",
    "name": "Cole Runolfsson",
  },
  Object {
    "avatar": Object {
      "color": "red",
      "image": Object {
        "path": "/images/avatars/w2.png",
        "url": "https://res.cloudinary.com/lucassus/image/upload/bookshelf/images/avatars/w2.png",
      },
    },
    "info": "Velit molestiae voluptates exercitationem reiciendis.",
    "name": "Oceane Franecki",
  },
]
`;
